common/state/state_kv.cpp

```diff
@@ -935,12 +1335,79 @@ void pstate::data_node_io::add_and_init_append_data_node()
     block_warehouse_.add_datablock_id(dn_id);
 }
 
+void pstate::data_node_io::consume_add_and_init_append_data_node()
+{
+    //consume remaining space (if any) in data node, for later collection
+    block_offset_t bo;
+    block_offset_for_appending(bo);
+    free_space_collector_.collect(bo, append_dn_->free_bytes());
+    append_dn_->consume_free_space(append_dn_->free_bytes());
+
+    add_and_init_append_data_node();
+}
+
 void pstate::data_node_io::add_and_init_append_data_node_cond(bool cond)
 {
     if (cond)
         pstate::data_node_io::add_and_init_append_data_node();
 }
 
+void pstate::data_node_io::consume_add_and_init_append_data_node_cond(bool cond)
+{
+    if (cond)
+        pstate::data_node_io::consume_add_and_init_append_data_node();
+}
+
+void pstate::data_node_io::write_across_data_nodes(const ByteArray& buffer, unsigned int write_from, const block_offset_t& bo_at)
+{
+    block_offset_t bo = bo_at;
+
+    unsigned int bytes_written, total_bytes_written = 0;
+
+    // start writing value
+    while(total_bytes_written < buffer.size())
+    {
+        data_node& dn = cache_retrieve(bo.block_num, false);
+        bytes_written = dn.write_at(buffer, total_bytes_written, bo);
+        cache_done(bo.block_num, true);
+
+        //increment written bytes and advance block offset
+        total_bytes_written += bytes_written;
+        data_node::advance_block_offset(bo, bytes_written);
+
+        //if we are appending and the block offset touches a new data node, make sure to append a new one to the list
+        add_and_init_append_data_node_cond(bo.block_num > block_warehouse_.last_appended_data_block_num_);
+    }
+}
+
+void pstate::data_node_io::read_across_data_nodes(const block_offset_t& bo_at, unsigned int length, ByteArray& out_buffer)
+{
+    block_offset_t bo = bo_at;
+
+    unsigned int bytes_read, total_bytes_read = 0;
+
+    // start reading value
+    while(total_bytes_read < length)
+    {
+
+        data_node& dn = cache_retrieve(bo.block_num, false);
+        try
+        {
+        bytes_read = dn.read_at(bo, length - total_bytes_read, out_buffer);
+        }
+        catch (std::exception& e)
+        {
+            SAFE_LOG(PDO_LOG_DEBUG, "read at call failed");
+            throw pdo::error::Error(PDO_ERR_RUNTIME, std::string("read at, call failed, ") + std::string(e.what()));
+        }
```

Comment:
why are you catching a general error but throwing a PDO error?

Reply:
changed